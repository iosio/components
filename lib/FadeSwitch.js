(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', './Fader'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('./Fader'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.Fader);
        global.FadeSwitch = mod.exports;
    }
})(this, function (exports, _react, _Fader) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.FadeSwitch = undefined;

    var _react2 = _interopRequireDefault(_react);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }

    var _createClass = function () {
        function defineProperties(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || false;
                descriptor.configurable = true;
                if ("value" in descriptor) descriptor.writable = true;
                Object.defineProperty(target, descriptor.key, descriptor);
            }
        }

        return function (Constructor, protoProps, staticProps) {
            if (protoProps) defineProperties(Constructor.prototype, protoProps);
            if (staticProps) defineProperties(Constructor, staticProps);
            return Constructor;
        };
    }();

    function _possibleConstructorReturn(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    }

    function _inherits(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    }

    var FadeSwitch = exports.FadeSwitch = function (_React$Component) {
        _inherits(FadeSwitch, _React$Component);

        function FadeSwitch(props) {
            _classCallCheck(this, FadeSwitch);

            var _this = _possibleConstructorReturn(this, (FadeSwitch.__proto__ || Object.getPrototypeOf(FadeSwitch)).call(this, props));

            _this.showView = function (view_num) {

                if (_this.is_mounted_) {

                    if (!_this.state.initial_load) {

                        _this.setState({ show: false });

                        _this.timeout = setTimeout(function () {
                            if (_this.is_mounted_) {
                                _this.setState({ view: view_num, show: true });
                            }
                        }, _this.duration);
                    } else {
                        _this.setState({ view: view_num, show: true, initial_load: false });
                    }
                }
            };

            _this.state = {
                show: false,
                view: 0,
                initial_load: true
            };
            _this.is_mounted_ = false;
            _this.timeout = false;
            _this.duration = props.duration ? props.duration : 175;
            return _this;
        }

        _createClass(FadeSwitch, [{
            key: 'componentDidMount',
            value: function componentDidMount() {
                this.is_mounted_ = true;
                this.showView(this.props.view);
            }
        }, {
            key: 'componentDidUpdate',
            value: function componentDidUpdate(prevProps) {
                if (prevProps.view !== this.props.view) {
                    this.showView(this.props.view);
                }
            }
        }, {
            key: 'componentWillUnmount',
            value: function componentWillUnmount() {
                this.is_mounted_ = false;
                clearTimeout(this.timeout);
            }
        }, {
            key: 'render',
            value: function render() {
                var _props = this.props,
                    View0 = _props.View0,
                    View1 = _props.View1,
                    className = _props.className,
                    fader_props = _props.fader_props,
                    style = _props.style;
                var _state = this.state,
                    show = _state.show,
                    view = _state.view;


                var fp_style = {};
                if (fader_props && fader_props.style) {
                    fp_style = fader_props.style;
                }
                var combined_style = Object.assign({}, style, fp_style);

                return _react2.default.createElement(
                    _Fader.Fader,
                    {
                        duration: this.duration,
                        className: className,
                        style: Object.assign({
                            height: '100%',
                            width: '100%',
                            position: 'relative'
                        }, combined_style),
                        show: show },
                    view === 0 ? View0 : View1
                );
            }
        }]);

        return FadeSwitch;
    }(_react2.default.Component);
});