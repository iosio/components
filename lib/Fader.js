(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', 'jss', 'jss-preset-default', 'classnames'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('jss'), require('jss-preset-default'), require('classnames'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.jss, global.jssPresetDefault, global.classnames);
        global.Fader = mod.exports;
    }
})(this, function (exports, _react, _jss, _jssPresetDefault, _classnames) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.Fader = undefined;

    var _react2 = _interopRequireDefault(_react);

    var _jss2 = _interopRequireDefault(_jss);

    var _jssPresetDefault2 = _interopRequireDefault(_jssPresetDefault);

    var _classnames2 = _interopRequireDefault(_classnames);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    function _objectWithoutProperties(obj, keys) {
        var target = {};

        for (var i in obj) {
            if (keys.indexOf(i) >= 0) continue;
            if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;
            target[i] = obj[i];
        }

        return target;
    }

    function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }

    var _createClass = function () {
        function defineProperties(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || false;
                descriptor.configurable = true;
                if ("value" in descriptor) descriptor.writable = true;
                Object.defineProperty(target, descriptor.key, descriptor);
            }
        }

        return function (Constructor, protoProps, staticProps) {
            if (protoProps) defineProperties(Constructor.prototype, protoProps);
            if (staticProps) defineProperties(Constructor, staticProps);
            return Constructor;
        };
    }();

    function _possibleConstructorReturn(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    }

    function _inherits(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    }

    _jss2.default.setup((0, _jssPresetDefault2.default)());
    var getStyles = function getStyles(styles) {
        var _jss$createStyleSheet = _jss2.default.createStyleSheet(styles).attach(),
            classes = _jss$createStyleSheet.classes;

        return classes;
    };

    var Fader = exports.Fader = function (_React$PureComponent) {
        _inherits(Fader, _React$PureComponent);

        function Fader(props) {
            _classCallCheck(this, Fader);

            var _this = _possibleConstructorReturn(this, (Fader.__proto__ || Object.getPrototypeOf(Fader)).call(this, props));

            var duration = props.duration;

            var value = duration ? duration : 175;
            var styles = {
                root: {
                    opacity: 0
                },
                trans: {
                    transition: 'opacity ' + value + 'ms ease-in-out'
                },
                show: {
                    opacity: 1
                }
            };
            _this.classes = getStyles(styles);
            return _this;
        }

        _createClass(Fader, [{
            key: 'render',
            value: function render() {
                var classes = this.classes;

                var _props = this.props,
                    className = _props.className,
                    style = _props.style,
                    show = _props.show,
                    children = _props.children,
                    ref = _props.ref,
                    rest = _objectWithoutProperties(_props, ['className', 'style', 'show', 'children', 'ref']);

                return _react2.default.createElement(
                    'div',
                    Object.assign({
                        ref: ref,
                        className: (0, _classnames2.default)(classes.root, classes.trans, show ? classes.show : null, className),
                        style: Object.assign({}, style)
                    }, rest),
                    children
                );
            }
        }]);

        return Fader;
    }(_react2.default.PureComponent);

    ;
});