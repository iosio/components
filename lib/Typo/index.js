(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', '@iosio/utils/lib/loaders/index', 'react-jss', './typography', 'classnames', '@iosio/utils/lib/string_manipulation'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('@iosio/utils/lib/loaders/index'), require('react-jss'), require('./typography'), require('classnames'), require('@iosio/utils/lib/string_manipulation'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.index, global.reactJss, global.typography, global.classnames, global.string_manipulation);
        global.index = mod.exports;
    }
})(this, function (exports, _react, _index, _reactJss, _typography, _classnames, _string_manipulation) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.buildTypo = undefined;

    var _react2 = _interopRequireDefault(_react);

    var _reactJss2 = _interopRequireDefault(_reactJss);

    var _classnames2 = _interopRequireDefault(_classnames);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }

    var _createClass = function () {
        function defineProperties(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || false;
                descriptor.configurable = true;
                if ("value" in descriptor) descriptor.writable = true;
                Object.defineProperty(target, descriptor.key, descriptor);
            }
        }

        return function (Constructor, protoProps, staticProps) {
            if (protoProps) defineProperties(Constructor.prototype, protoProps);
            if (staticProps) defineProperties(Constructor, staticProps);
            return Constructor;
        };
    }();

    function _possibleConstructorReturn(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    }

    function _inherits(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    }

    function _objectWithoutProperties(obj, keys) {
        var target = {};

        for (var i in obj) {
            if (keys.indexOf(i) >= 0) continue;
            if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;
            target[i] = obj[i];
        }

        return target;
    }

    var buildTypo = function buildTypo(config, cb) {
        var load_link = config.load_link,
            hierarchy = config.hierarchy,
            measure = config.measure,
            hierarchy_base_styles = config.hierarchy_base_styles,
            global_styles = config.global_styles,
            fontFamily = config.fontFamily,
            exclude_hierarchy_defaults = config.exclude_hierarchy_defaults,
            font_wights = config.font_wights,
            text_color = config.text_color,
            line_height_adjust = config.line_height_adjust,
            aux_styles = config.aux_styles;


        if (!line_height_adjust) {
            line_height_adjust = 2;
        }

        var aux = {
            _active: {
                color: '#1fb5fa',
                pointerEvents: 'auto'
            },
            _disabled: {
                opacity: 0.4,
                pointerEvents: 'none'
            },
            _error: {
                color: 'red'
            },
            _contrast: {
                '&:not($_error)': {
                    color: '#fff'
                }
            }
        };

        if (!aux_styles) {
            aux_styles = aux;
        } else {
            aux_styles = Object.assign({}, aux, aux_styles);
        }

        var color = {
            'default': '#242b36',
            darker: '#171a23',
            lighter: '#354051'
        };

        if (!text_color) {
            text_color = color;
        } else {
            text_color = Object.assign({}, color, text_color);
        }

        var weights = {
            thin: 100,
            light: 300,
            regular: 400,
            medium: 500,
            bold: 700,
            black: 900
        };

        if (!font_wights) {
            font_wights = weights;
        } else {
            font_wights = Object.assign({}, weights, font_wights);
        }

        var divides = {
            display4: 1,
            display3: 2,
            display2: 2.4,
            display1: 3.3,
            headline: 4.6,
            title: 5.3,
            subheading: 7,
            body2: 8,
            body1: 8,
            body_muted: 8,
            button: 8,
            caption: 10
        };

        if (!fontFamily) {
            fontFamily = '"Roboto", "Helvetica", "Arial", sans-serif';
        }

        var typography = (0, _typography.typography)({ weights: font_wights, text_color: text_color, fontFamily: fontFamily });

        if (load_link && load_link.url) {
            (0, _index.loadLink)(load_link.url, function () {
                return load_link.cb && load_link.cb();
            });
        }

        if (global_styles) {
            typography['@global'] = Object.assign({
                html: {
                    WebkitFontSmoothing: 'antialiased', // Antialiasing.
                    MozOsxFontSmoothing: 'grayscale' // Antialiasing.
                }
            }, global_styles);
        } else {
            typography['@global'] = {
                html: {
                    WebkitFontSmoothing: 'antialiased', // Antialiasing.
                    MozOsxFontSmoothing: 'grayscale' // Antialiasing.
                }
            };
        }

        if (!measure) {
            measure = 112;
        }

        if (!hierarchy) {

            hierarchy = {
                display4: {},
                display3: {},
                display2: {},
                display1: {},
                headline: {},
                title: {},
                subheading: {},
                body2: {},
                body1: {},
                body_muted: {},
                button: {},
                caption: {}
            };
        }

        if (!hierarchy_base_styles) {
            hierarchy_base_styles = {};
        }

        Object.keys(hierarchy).forEach(function (key) {
            var _hierarchy$key = hierarchy[key],
                font_measure_divide = _hierarchy$key.font_measure_divide,
                rest = _objectWithoutProperties(_hierarchy$key, ['font_measure_divide']);

            var fontSize = void 0;
            if (font_measure_divide) {
                fontSize = Math.floor(measure / font_measure_divide);
            } else {
                fontSize = Math.floor(measure / divides[key]);
            }

            var defaults = {};
            if (!exclude_hierarchy_defaults && typography[key]) {
                defaults = typography[key];
            }

            typography[key] = Object.assign({}, defaults, hierarchy_base_styles, {
                fontSize: fontSize,
                lineHeight: fontSize + line_height_adjust + 'px'
            }, rest);
        });

        typography = Object.assign({}, typography, aux_styles);

        var Typo = function (_React$Component) {
            _inherits(Typo, _React$Component);

            function Typo() {
                _classCallCheck(this, Typo);

                return _possibleConstructorReturn(this, (Typo.__proto__ || Object.getPrototypeOf(Typo)).apply(this, arguments));
            }

            _createClass(Typo, [{
                key: 'render',
                value: function render() {
                    var _props = this.props,
                        classes = _props.classes,
                        hierarchy = _props.hierarchy,
                        block = _props.block,
                        children = _props.children,
                        text = _props.text,
                        className = _props.className,
                        style = _props.style,
                        elem = _props.elem,
                        replace = _props.replace,
                        ellipsis = _props.ellipsis,
                        cap = _props.cap,
                        maxChar = _props.maxChar,
                        align = _props.align,
                        space = _props.space,
                        select_none = _props.select_none,
                        rest = _objectWithoutProperties(_props, ['classes', 'hierarchy', 'block', 'children', 'text', 'className', 'style', 'elem', 'replace', 'ellipsis', 'cap', 'maxChar', 'align', 'space', 'select_none']);

                    var combined_classes = [className];

                    var others = {};

                    Object.keys(rest).forEach(function (key) {

                        if (classes[key] && rest[key] === true && key.charAt(0) === '_') {
                            combined_classes.push(classes[key]);
                        } else if (key.charAt(0) !== '_') {
                            others[key] = rest[key];
                        }
                    });

                    if (classes[hierarchy]) {
                        combined_classes.push(classes[hierarchy]);
                    }

                    var Component = elem ? elem : 'span';
                    var _text = text ? text : children ? children : '';
                    var show_ellipsis = false;

                    var combined_style = {};

                    if (style) {
                        combined_style = style;
                    }

                    if (space) {
                        combined_style.marginLeft = 5;
                        combined_style.marginRight = 5;
                    }

                    if (select_none) {
                        combined_style.userSelect = 'none';
                    }

                    if (block) {
                        combined_style.display = 'block';
                    }
                    if (align) {
                        combined_style.textAlign = align;
                    }

                    if (_text) {
                        if (maxChar) {
                            _text = _text.substring(0, maxChar);
                        }

                        if (ellipsis === 'singleLine') {
                            combined_style.whiteSpace = 'nowrap';
                            combined_style.overflow = 'hidden';
                            combined_style.textOverflow = 'ellipsis';
                            combined_style.display = 'block';
                        } else if (ellipsis === 'atMaxChar' && maxChar) {
                            show_ellipsis = _text.length >= maxChar;
                        } else if (ellipsis === true) {
                            show_ellipsis = true;
                        }

                        if (replace) {
                            _text = (0, _string_manipulation.textReplacer)(_text, replace);
                        }
                        if (cap) {
                            _text = (0, _string_manipulation.capitalize)(_text, cap);
                        }
                    }

                    // console.log(combined_classes);

                    return _react2.default.createElement(
                        Component,
                        Object.assign({
                            className: _classnames2.default.apply(undefined, combined_classes),
                            style: combined_style
                        }, others),
                        _text,
                        show_ellipsis && _react2.default.createElement(
                            'span',
                            { style: { fontSize: '18px' } },
                            ' ...'
                        )
                    );
                }
            }]);

            return Typo;
        }(_react2.default.Component);

        return {
            fontFamily: fontFamily,
            typography: typography,
            hierarchy: hierarchy,
            Typo: (0, _reactJss2.default)(typography)(Typo)
        };
    };
    exports.buildTypo = buildTypo;
});