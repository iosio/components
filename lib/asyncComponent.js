(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', './FadeSwitch'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('./FadeSwitch'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.FadeSwitch);
        global.asyncComponent = mod.exports;
    }
})(this, function (exports, _react, _FadeSwitch) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.asyncComponent = undefined;

    var _react2 = _interopRequireDefault(_react);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }

    var _createClass = function () {
        function defineProperties(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || false;
                descriptor.configurable = true;
                if ("value" in descriptor) descriptor.writable = true;
                Object.defineProperty(target, descriptor.key, descriptor);
            }
        }

        return function (Constructor, protoProps, staticProps) {
            if (protoProps) defineProperties(Constructor.prototype, protoProps);
            if (staticProps) defineProperties(Constructor, staticProps);
            return Constructor;
        };
    }();

    function _possibleConstructorReturn(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    }

    function _inherits(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    }

    var asyncComponent = exports.asyncComponent = function asyncComponent(getComponent, Loader, duration, fadeSwitch_props, fader_props) {
        var _class, _temp;

        return _temp = _class = function (_React$Component) {
            _inherits(AsyncComponent, _React$Component);

            function AsyncComponent() {
                _classCallCheck(this, AsyncComponent);

                var _this = _possibleConstructorReturn(this, (AsyncComponent.__proto__ || Object.getPrototypeOf(AsyncComponent)).call(this));

                _this.state = { Component: AsyncComponent.Component };

                if (!_this.state.Component) {
                    getComponent().then(function (Component) {
                        Component = Component && Component.__esModule ? Component.default : Component;
                        AsyncComponent.Component = Component;
                        _this.setState({ Component: Component });
                    });
                }
                return _this;
            }

            _createClass(AsyncComponent, [{
                key: 'render',
                value: function render() {
                    var Component = this.state.Component;


                    return _react2.default.createElement(
                        _react2.default.Fragment,
                        null,
                        duration > 0 ? _react2.default.createElement(_FadeSwitch.FadeSwitch, Object.assign({}, fadeSwitch_props, {
                            fader_props: fader_props,
                            duration: duration
                        }, this.props, {
                            style: { height: '100%', width: '100%' },
                            view: !!Component ? 1 : 0,
                            View0: Loader ? _react2.default.createElement(Loader, null) : null,
                            View1: Component ? _react2.default.createElement(Component, this.props) : null
                        })) : Component ? _react2.default.createElement(Component, this.props) : null
                    );
                }
            }]);

            return AsyncComponent;
        }(_react2.default.Component), _class.Component = null, _temp;
    };
});