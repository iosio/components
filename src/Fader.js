import React from 'react';
export class Fader extends React.PureComponent {
    render() {
        const { className, style, show, children, duration, ref, ...rest } = this.props;
        const value = duration ? duration : 175;
        return (
            <div
                id="Fader"
                ref={ref}
                className={className}
                style={{
                    ...style,
                    opacity: show ? 1: 0,
                    transition: `opacity ${value}ms ease-in-out`
                }}
                {...rest}>

                {children}
            </div>
        );
    }
}









