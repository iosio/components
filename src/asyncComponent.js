import React from 'react';
import {FadeSwitch} from "./FadeSwitch";

export const asyncComponent = (getComponent, Loader, duration, fadeSwitch_props, fader_props) => {
    return class AsyncComponent extends React.Component {
        constructor() {
            super();
            if (!this.state.Component) {
                getComponent().then(Component => {
                    Component = Component && Component.__esModule ? Component.default : Component;
                    AsyncComponent.Component = Component;
                    this.setState({Component});
                });
            }
        }

        static Component = null;
        state = {Component: AsyncComponent.Component};

        render() {
            const {Component} = this.state;

            return (
                <React.Fragment>
                    {Loader && duration > 0 ? <FadeSwitch
                            {...fadeSwitch_props}
                            fader_props={fader_props}
                            duration={duration}
                            {...this.props}
                            style={{height: '100%', width: '100%'}}
                            view={!!Component ? 1 : 0}
                            View0={Loader ? <Loader/> : null}
                            View1={Component ? <Component {...this.props}/> : null}
                        />
                        :

                        (Component ? <Component {...this.props}/> : null)
                    }
                </React.Fragment>
            );
        }
    }
};


