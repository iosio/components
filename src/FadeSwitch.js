import React from 'react';

import {Fader} from "./Fader";

export class FadeSwitch extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            show: false,
            view: 0,
            initial_load: true,
        };
        this.is_mounted_ = false;
        this.timeout = false;
        this.duration = props.duration || props.duration === 0 ? props.duration : 175;
    }

    componentDidMount() {
        this.is_mounted_ = true;
        this.showView(this.props.view);
    }

    showView = (view_num) => {

        if (this.is_mounted_) {

            if (!this.state.initial_load) {

                this.setState({show: false});

                this.timeout = setTimeout(() => {
                    if (this.is_mounted_) {
                        this.setState({view: view_num, show: true,});
                    }
                }, this.duration);

            } else {
                this.setState({view: view_num, show: true, initial_load: false});
            }
        }
    };

    componentDidUpdate(prevProps) {
        if (prevProps.view !== this.props.view) {
            this.showView(this.props.view);
        }
    }

    componentWillUnmount() {
        this.is_mounted_ = false;
        clearTimeout(this.timeout);
    }

    render() {
        const {
            View0,
            View1,
            className,
            fader_props,
            style
        } = this.props;

        const {
            show,
            view,
        } = this.state;

        let fp_style = {};
        if (fader_props && fader_props.style) {
            fp_style = fader_props.style;
        }
        const combined_style = {...style, ...fp_style};

        return (
            <Fader
                duration={this.duration}
                className={className}
                style={{
                    height: '100%',
                    width: '100%',
                    position: 'relative',
                    ...combined_style,
                }}
                show={show}>

                {view === 0 ? View0 : View1}
            </Fader>

        );
    };
}
