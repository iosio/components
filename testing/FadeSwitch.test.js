import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {FadeSwitch} from "../src/FadeSwitch";
import {StoreModule} from "@iosio/capsule/lib/redux-assist";

Enzyme.configure({adapter: new Adapter()});

import {mount, shallow, render} from 'enzyme';

const {reduxAssist, store, connect, Provider} = new StoreModule();

const state = reduxAssist('view', {show: 0});

const wait = (time) => {
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve()
        }, time);
    })
};


class FadeSwitchWithConnect extends React.Component {

    render() {
        const {show} = this.props;

        return (
            <FadeSwitch
                duration={1}
                view={show}
                View0={
                    <div className="view-0">
                        I'm view 0!
                    </div>
                }

                View1={
                    <div className="view-1">
                        I'm view 1!
                    </div>
                }
            />
        )
    }
}

FadeSwitchWithConnect = connect((state) => ({
    show: state.view.show,
}), null)(FadeSwitchWithConnect);


describe('FadeSwitch Component', () => {

    it('should match snapshot when view = 0', () => {

        const tree = renderer.create(
            <FadeSwitch
                view={0}
                View0={
                    <div>
                        I'm view 0!
                    </div>
                }

                View1={
                    <div>
                        I'm view 1!
                    </div>
                }
            />
        );

        expect(tree).toMatchSnapshot();

    });

    it('should match snapshot when view = 1', () => {

        const tree = renderer.create(
            <FadeSwitch
                view={1}
                View0={
                    <div>
                        I'm view 0!
                    </div>
                }

                View1={
                    <div>
                        I'm view 1!
                    </div>
                }
            />
        );

        expect(tree).toMatchSnapshot();
    });

    it('should match snapshot when duration is configured', () => {

        const tree = renderer.create(
            <FadeSwitch
                duration={500}
                view={1}
                View0={
                    <div>
                        I'm view 0!
                    </div>
                }

                View1={
                    <div>
                        I'm view 1!
                    </div>
                }
            />
        );

        expect(tree).toMatchSnapshot();
    });


    describe('it should update when new props are passed', () => {

        const tree = renderer.create(
            <Provider>
                <FadeSwitchWithConnect/>
            </Provider>
        );

        it('view (0)', () => {

            expect(tree).toMatchSnapshot();

        });

        it('view (1)', () => {
            expect.assertions(1);
            state.view.set.show(1);
            return wait(10).then(() => {
                expect(tree).toMatchSnapshot();
            })

        });


    });

    // describe('(enzyme) - it should update when new props are passed', () => {
    //
    //     const wrapper = mount(
    //         <Provider>
    //             <FadeSwitchWithConnect/>
    //         </Provider>
    //     );
    //
    //
    //     it('view (0)', () => {
    //
    //        console.log(wrapper.instance())
    //
    //         expect.assertions(1);
    //         return wait(10).then(()=>{
    //             expect(wrapper.find('.view-1').length).toBe(1);
    //         })
    //     });
    //
    //     // it('view (1)', () => {
    //     //     expect.assertions(1);
    //     //     state.view.set.show(1);
    //     //     return wait(10).then(()=>{
    //     //         console.log(tree.find('.view-1').exists());
    //     //         expect(tree.find('.view-1').length).toBe(1);
    //     //     })
    //     //
    //     // });
    //
    //
    // })


});