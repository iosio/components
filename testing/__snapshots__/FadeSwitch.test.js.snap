// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FadeSwitch Component it should update when new props are passed view (0) 1`] = `
<div
  id="Fader"
  style={
    Object {
      "height": "100%",
      "opacity": 1,
      "position": "relative",
      "transition": "opacity 1ms ease-in-out",
      "width": "100%",
    }
  }
>
  <div
    className="view-0"
  >
    I'm view 0!
  </div>
</div>
`;

exports[`FadeSwitch Component it should update when new props are passed view (1) 1`] = `
<div
  id="Fader"
  style={
    Object {
      "height": "100%",
      "opacity": 1,
      "position": "relative",
      "transition": "opacity 1ms ease-in-out",
      "width": "100%",
    }
  }
>
  <div
    className="view-1"
  >
    I'm view 1!
  </div>
</div>
`;

exports[`FadeSwitch Component should match snapshot when duration is configured 1`] = `
<div
  id="Fader"
  style={
    Object {
      "height": "100%",
      "opacity": 1,
      "position": "relative",
      "transition": "opacity 500ms ease-in-out",
      "width": "100%",
    }
  }
>
  <div>
    I'm view 1!
  </div>
</div>
`;

exports[`FadeSwitch Component should match snapshot when view = 0 1`] = `
<div
  id="Fader"
  style={
    Object {
      "height": "100%",
      "opacity": 1,
      "position": "relative",
      "transition": "opacity 175ms ease-in-out",
      "width": "100%",
    }
  }
>
  <div>
    I'm view 0!
  </div>
</div>
`;

exports[`FadeSwitch Component should match snapshot when view = 1 1`] = `
<div
  id="Fader"
  style={
    Object {
      "height": "100%",
      "opacity": 1,
      "position": "relative",
      "transition": "opacity 175ms ease-in-out",
      "width": "100%",
    }
  }
>
  <div>
    I'm view 1!
  </div>
</div>
`;
